tests/it/arrow.rs

```diff
@@ -0,0 +1,350 @@
+use arrow2::array::*;
+use arrow2::bitmap::Bitmap;
+use arrow2::datatypes::{DataType, Field, IntegerType, TimeUnit, UnionMode};
+use arrow2::offset::Offsets;
+use arrow_array::ArrayRef;
+use arrow_data::ArrayDataBuilder;
+use proptest::num::i32;
+
+fn test_arrow2_roundtrip(array: &dyn arrow_array::Array) {
+    let arrow2 = Box::<dyn Array>::from(array);
+    assert_eq!(arrow2.len(), array.len());
+
+    let back = ArrayRef::from(arrow2);
+    assert_eq!(back.len(), array.len());
+
+    match array.data_type() {
+        d @ arrow_schema::DataType::Union(_, _, arrow_schema::UnionMode::Sparse) => {
+            // Temporary workaround https://github.com/apache/arrow-rs/issues/4044
+            let data = array.to_data();
+            let type_ids = data.buffers()[0].slice_with_length(data.offset(), data.len());
+            let child_data = data
+                .child_data()
+                .iter()
+                .map(|x| x.slice(data.offset(), data.len()))
+                .collect();
+
+            let data = ArrayDataBuilder::new(d.clone())
+                .len(data.len())
+                .buffers(vec![type_ids])
+                .child_data(child_data)
+                .build()
+                .unwrap();
+
+            assert_eq!(back.to_data(), data);
+        }
+        _ => assert_eq!(array, back.as_ref()),
+    }
+    assert_eq!(array.data_type(), back.data_type());
+}
+
+fn test_arrow_roundtrip(array: &dyn Array) {
+    let arrow = ArrayRef::from(array);
+    assert_eq!(arrow.len(), array.len());
+
+    let back = Box::<dyn Array>::from(arrow);
+    assert_eq!(back.len(), array.len());
+
+    assert_eq!(array, back.as_ref());
+    assert_eq!(array.data_type(), back.data_type());
+}
+
+fn test_conversion(array: &dyn Array) {
+    test_arrow_roundtrip(array);
+    let to_arrow = ArrayRef::from(array);
+    test_arrow2_roundtrip(to_arrow.as_ref());
+
+    if !array.is_empty() {
+        let sliced = array.sliced(1, array.len() - 1);
+        test_arrow_roundtrip(sliced.as_ref());
+
+        let sliced = to_arrow.slice(1, array.len() - 1);
+        test_arrow2_roundtrip(sliced.as_ref());
+
+        let sliced = array.sliced(0, array.len() - 1);
+        test_arrow_roundtrip(sliced.as_ref());
+
+        let sliced = to_arrow.slice(0, array.len() - 1);
+        test_arrow2_roundtrip(sliced.as_ref());
+    }
+
+    if array.len() > 2 {
+        let sliced = array.sliced(1, array.len() - 2);
+        test_arrow_roundtrip(sliced.as_ref());
+
+        let sliced = to_arrow.slice(1, array.len() - 2);
+        test_arrow2_roundtrip(sliced.as_ref());
+    }
+}
+
+#[test]
+fn test_primitive() {
+    let data_type = DataType::Int32;
+    let array = PrimitiveArray::new(data_type, vec![1, 2, 3].into(), None);
```

Comment:
The only way I can think of to make these tests better is to manually construct the arrow-rs array as well (otherwise there is just one original source of truth). However I think these tests are very impressive overall and don't actually need anything more